C front-end: fix processing of alignment and packing attributes

tautschnig opened the pull request:
Packing needs to be evaluated as part of the definition of the type while alignment applies also when using a type. The position of the attribute also requires extra care as some positions are accepted by GCC (and warned about by Clang) while actually being ignored.

Fixes: #8443

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
C front-end: fix processing of alignment and packing attributes

Packing needs to be evaluated as part of the definition of the type
while alignment applies also when using a type. The position of the
attribute also requires extra care as some positions are accepted by GCC
(and warned about by Clang) while actually being ignored.

Fixes: #8443
Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
C front-end: fix processing of alignment and packing attributes

Packing needs to be evaluated as part of the definition of the type
while alignment applies also when using a type. The position of the
attribute also requires extra care as some positions are accepted by GCC
(and warned about by Clang) while actually being ignored.

The alignment validator now rejects non-positive and non-power-of-two
values via a small is_power_of_two helper (mp_integer does not provide
bitwise operators, so this uses floorPow2). The alignment-combining
helper is renamed to combine_alignments and documents/asserts its
power-of-two precondition (under which the maximum equals the least
common multiple). The grammar action that drops ignored packed/aligned
attributes is factored into a strip_ignored_attributes helper so the
rule reads like grammar.

The regression test gains _Alignof assertions (so alignments that don't
change size are actually checked), packed/aligned union cases, a
forward-declared-then-completed struct that exercises the
alignment-combining path, a single-ignored-attribute-plus-qualifier case,
and a direct reduction of #8443 (a packed-struct typedef used inside an
outer struct carrying an alignment attribute). A new negative test,
gcc_attributes16_align_error, pins the "not a positive power of 2" error
path.

Fixes: #8443
Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening commented on src/ansi-c/c_typecheck_type.cpp:
```diff
@@ -72,6 +72,14 @@ void c_typecheck_baset::typecheck_type(typet &type)
+        power(2, align_int->floorPow2()) != *align_int)
+      {
+        throw errort().with_location(type.source_location())
+          << "alignment is not a positive power of 2";
```
Is that a requirement in the standard?

tautschnig replied:
C11 6.2.8 "Alignment of objects" includes the following sentence: "Every valid alignment value shall be a nonnegative integral power of two."

tautschnig committed:
C front-end: factor out alignment_of helper

Replace the five repeated static_cast<const exprt &>(t.find(ID_C_alignment))
casts in c_typecheck_type.cpp with a small alignment_of(const typet &)
helper. This centralises the cast (and the assumption that exprt adds no
data members to irept, so the cast is safe) in one documented place.

This is a pure refactoring with no change in behaviour.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
C front-end: pad union size up to its alignment

A union's size must be a multiple of its alignment, including when the
alignment comes from an explicit aligned() attribute on a packed union.
Previously add_padding(union_typet) discarded any explicit alignment for
packed unions (rounding the size up to a single byte only), so e.g.

  union __attribute__((packed)) __attribute__((aligned(16))) u
  { uint8_t c; int32_t i; };

had size 4 instead of 16. The alignment is now computed as the maximum
member alignment (one byte if packed), raised by an explicit aligned()
attribute regardless of packing, matching GCC and Clang.

Adds regression test union_packed_aligned.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
C front-end: context-dependent member alignment

The alignment of a struct/union member depends on whether the enclosing
aggregate is packed, which a context-free alignment() could not capture:
the same type (e.g. int aligned(2)) must align to 4 inside a non-packed
struct (an aligned() attribute can only increase alignment there) but to 2
inside a packed struct (where it is honoured verbatim).

alignment() now derives an aggregate's alignment from its members' layout
alignments, computed in the aggregate's packing context via a new
member_layout_alignment helper:
- packed field: the explicit alignment verbatim, otherwise one byte;
- non-packed field: max(explicit, natural) -- an aligned() request smaller
  than the natural alignment is ignored.
An explicit alignment on the aggregate itself can only raise the result,
while on a non-aggregate (e.g. a typedef) it is still taken verbatim.
Padding members no longer contribute to an aggregate's alignment (they
previously polluted _Alignof of e.g. a union with an array member).

add_padding_gcc uses the same per-member computation for member offsets,
and a packed struct is now padded up to the alignment forced by an
explicitly-aligned member (it previously skipped end-of-struct padding for
any packed struct). add_padding(union_typet) is simplified to rely on the
corrected alignment().

This makes sizeof/_Alignof match GCC and Clang for member-level and
struct/union-level aligned()/packed combinations. Adds regression test
struct_member_alignment.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
fixup! C front-end: context-dependent member alignment

tautschnig committed:
C front-end: honour reduced-alignment typedefs used as struct members

GCC, Clang and Intel icx allow __attribute__((aligned(n))) on a typedef to
*reduce* a type's alignment, and that reduced alignment is kept when the
typedef is used as a struct member (the member, and hence the enclosing
struct, is aligned to n). An alignment attribute on an object, field or tag
declaration, by contrast, can only *increase* the alignment; a smaller
request is ignored.

The front-end stored both kinds identically as ID_C_alignment, so padding
could not tell them apart: member_layout_alignment treated every member
alignment as increase-only and over-aligned a reduced-alignment typedef
member, inflating the enclosing struct. ID_C_typedef does not disambiguate
either, as it is also present when a reducing attribute is attached to the
declarator of a plain-typedef'd field.

Record the provenance with a new ID_C_alignment_increase_only flag, set by
ansi_c_convert_typet::set_attributes whenever an alignment attribute is
applied to a non-typedef declaration (the only place that distinction is
known). The flag is preserved through type checking -- the already-typechecked
rebuild and typedef resolution, where a typedef's own alignment is left
verbatim. padding then applies the matching rule: packing caps the alignment
at min(n, natural); an increase-only attribute only raises it; a type-level
alignment is taken verbatim and may reduce.

This promotes regression/ansi-c/struct_member_alignment_typedef from KNOWNBUG
to CORE.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
